source/common/config/grpc_subscription_impl.h

```diff
@@ -19,6 +19,7 @@ class GrpcSubscriptionImpl : public Config::Subscription<ResourceType> {
                        const Protobuf::MethodDescriptor& service_method, SubscriptionStats stats)
       : grpc_mux_(node, std::move(async_client), dispatcher, service_method, random),
         grpc_mux_subscription_(grpc_mux_, stats) {}
+  ~GrpcSubscriptionImpl() { grpc_mux_.noMoreRequestSending(); }
```

Comment:
This doesn't seem correct; `grpc_mux_` might be an ADS connection. When I delete an individual subscriber, e.g. EDS, I shouldn't permanently disable the muxer. It seems like what we had before is the right thing, the gRPC subscription should relinquish its watch by deleting it, and then via RAII it cleans up its state in the gRPC mux. But, from the issue, there is some bug still. Can you deep dive?

Reply:
GrpcSubscriptionImpl owns grpc_mux_.   Look at line 39 of this file.
At its destructor,  grpc_mux_ is about to be deleted.    It is OK to mark all its watchers; not to send request when a watcher is removed.

Reply:
Oh, right, was confusing this with `GrpcMuxSubscriptionImpl`. It's still not right IMHO, since you're reaching in and dealing with internals in a way that breaks abstractions. The correct solution should clean up via RAII of watchers.

Reply:
I had a look at the code a bit more. I think even with your change, it's still think there is a race to worry about. The completion thread should still be live when the client is deleted, as that lives in TLS (are you seeing https://github.com/envoyproxy/envoy/blob/master/source/common/grpc/google_async_client_impl.cc#L19 invoked?). The async client should have reset the stream on its way out. I wonder if we have an issue in https://github.com/envoyproxy/envoy/blob/master/source/common/grpc/google_async_client_impl.cc#L363?

Can you provide a full trace level log and backtrace with line numbers? Thanks.

Reply:
(I'm following up with @qiwzhang on IM offline, it looks like SDS still has an active gRPC subscription after TLS shutdown has begun, this is dangerous and we need to figure out a way to avoid that)